### ISSUE TITLE
[Forms / iOS] Adding multiple layers causes app freeze

### ISSUE THREAD

**Issue by Sebastian1989101** [Mapsui/Mapsui#683]
As I started to redo my Breath Companion app with Mapsui 2.0, I encounter an issue where the whole map freezes in the process of adding the layers to the map. I think this comes from executing all events one by one because there is no way to add/remove/replace multiple layers at once. This was my code to add the layers (layer 0 is the tile layer so this one is skipped on removal): 

```
var entries = MapEntries.ToArray();
var layers = new List<ILayer>();

if (MapLabels != null && MapLabels.Any())
	layers.AddRange(GenerateTextLayers(entries, MapLabels.ToArray()));

if (MapLines != null && MapLines.Any())
	layers.AddRange(GenerateLineLayers(MapLines.ToArray()));

foreach (var marker in MapMarkers)
	layers.AddRange(GenerateLayers(entries, marker));

while (Map.Layers.Count > 1)
	Map.Layers.Remove(Map.Layers[1]);
	
foreach (var layer in layers)
	Map.Layers.Add(layer);
```

The app has ~800 layers which is no problem on Android and neither was it with Mapsui 1.x. I think a `AddRange` and/or `RemoveRange` would fix this pretty easy. Otherwise it would help if iOS handles this the same as Android if this is possible.

**Comment by pauldendulk** [comment]
Does the app recover from the app freeze? 

For each layer added there is a OnLayerAdded event. And AddRange could reduce the number of events, but I would also change the OnLayerAdded to take a list of layer as argument.

The OnLayerAdded should not cause major changes so maybe it is also worthwhile to investigate why it takes this long (if it recovers at all).

**Comment by Sebastian1989101** [comment]
@pauldendulk Yes it does recover but it takes a huge amount of time. Measured by my Breath Companion app about 10-15 minutes on a clean iPhone X. Currently I implemented a async loop that adds one layer at a time and always waits (depending on the layer) 2-5ms. This prevents the freeze but makes the layers appear slowly (but I guess that is better than a freeze).

**Comment by charlenni** [comment]
800 are a huge number. Perhaps a repaint appears after adding each a layer?

**Comment by Sebastian1989101** [comment]
@charlenni I think that is exactly what happens. With the xamarin profiler I can see that some rendering task on the gpu is always on 100% while the freeze is happening and the cpu is also at about 30-40% at the same moment.

**Comment by pauldendulk** [comment]
There should be just an 'invalidate' call which would not cause a repaint on every LayerAdded event, but perhaps something goes wrong there. Since this is in forms and not a problem in iOS the problem might be in a dependency.

**Comment by pauldendulk** [comment]
hmm, perhaps it is not the problem with redraw but the repeated call to RefreshData. This is the stacktrace:

![image](https://user-images.githubusercontent.com/963462/65360420-2d527480-dc00-11e9-9d85-d494941dbdcd.png)

**Comment by pauldendulk** [comment]
Still, why would there be a difference with iOS.

**Comment by charlenni** [comment]
Do you have a sample for this?

**Comment by pauldendulk** [comment]
@Sebastian1989101 What was the reason you have this many layers? I think you explained somewhere but I forgot where. In general I would say the app is not intended for this. With respect to performance I will be more inclined to optimize for many feature than for many layers.

**Comment by Sebastian1989101** [comment]
@pauldendulk I use so many layers in this app to have multiple icons and each one grouped by it's icon type (so 1 layer) so I can show/hide the layer at runtime. Even with so many layers it runs pretty well even on very old devices like the iPhone 4S. There is just this freeze if I add all at once with Mapsui 2.0 which was not there in 1.x. Adding the layers step by step works fine too, it just causes the app to have 2-3 seconds each time the icon filter changes (or the tile layer) to recreate (or add/remove) the layers.

**Comment by pauldendulk** [comment]
I assume this was solved.

**Comment by janusw** [comment]
I guess the issue was finally solved by #1185?

**Comment by Sebastian1989101** [comment]
@janusw yes, this would be solved by #1185 - at least as long as the mass edit methods are used instead of the single ones.

**Comment by janusw** [comment]
Right :)

I just wanted to make a connection between the issue and the PR for future reference (since the PR did not mention this at all).

**Comment by charlenni** [comment]
Could be, that it is solved with #1192 too

**Comment by janusw** [comment]
> Could be, that it is solved with #1192 too

Possibly mitigated a bit. However, I also saw similar problems recently when adding multiple layers at once (with the changes from #1192, but without using the new mass-edit methods).
